### ISSUE TITLE
[afl] Debugging AFL check build failures

### ISSUE THREAD

**Issue by asraa** [google/oss-fuzz#4707]
Hi!

I have a quick question about AFL's check_build. I am getting errors that some testcases are taking too long and resulting in a timeout. Are there any tips on debugging? I might be able to tweak testcases to be shorter, but I have no clue which testcases they are from the logs:

```
BAD BUILD: fuzzing /tmp/not-out/json_fuzz_test with afl-fuzz failed.
/tmp/not-out/afl-fuzz -m none -i /tmp/json_fuzz_test_corpus -o /tmp/json_fuzz_test_afl_address_out /tmp/not-out/json_fuzz_test
afl-fuzz 2.57b by <lcamtuf@google.com>
[+] Disabling the UI because AFL_NO_UI is set.
[+] You have 12 CPU cores and 38 runnable tasks (utilization: 317%).
[!] WARNING: System under apparent load, performance may be spotty.
[!] WARNING: Not binding to a CPU core (AFL_NO_AFFINITY set).
[*] Checking core_pattern...
[*] Setting up output directories...
[+] Output directory exists but deemed OK to reuse.
[*] Deleting old session data...
[+] Output dir cleanup successful.
[*] Scanning '/tmp/json_fuzz_test_corpus'...
[+] No auto-generated dictionary tokens to reuse.
[*] Creating hard links for all input files...
[*] Validating target binary...
[+] Persistent mode binary detected.
[+] Deferred forkserver binary detected.
[*] Attempting dry run with 'id:000000,orig:input'...
[*] Spinning up the fork server...
[+] All right - fork server is up.

[-] The program took more than 1000 ms to process one of the initial test cases.
    This is bad news; raising the limit with the -t option is possible, but
    will probably make the fuzzing process extremely slow.

    If this test case is just a fluke, the other option is to just avoid it
    altogether, and find one that is less of a CPU hog.

[-] PROGRAM ABORT : Test case 'id:000000,orig:input' results in a timeout
         Location : perform_dry_run(), afl-fuzz.c:2806
```

How can I identify/get more info on `id:000000,orig:input`? I can build and run my fuzzers locally with libfuzzer, they aren't anywhere close to 1000 ms to run.

Thank you!
Asra

**Comment by jonathanmetzman** [comment]
I'll try to reproduce this.
Does this target use threads by any chance?

**Comment by fatimahkj** [comment]
> I'll try to reproduce this.
> Does this target use threads by any chance?

Hi Jonathan, 
I'm having the same issue but my test cases are very simple. 
and the target is a multithreaded application. 

Is there any way to solve the problem?

**Comment by jonathanmetzman** [comment]
@asraa Sorry I forgot about this. I'll try to repro this again.

**Comment by asraa** [comment]
No worries, I did too over the short week. The target shouldn't be using threads

**Comment by jonathanmetzman** [comment]
Thanks!

**Comment by jonathanmetzman** [comment]
> No worries, I did too over the short week. The target shouldn't be using threads

OK I can reproduce this with `check_build`. The contents of that file are just "input\n".
Notably, I can't repro this just using afl-fuzz. This makes me think that the problem is the parallelism. I bet because bad_build_check runs so many fuzzers in parallel it slows down AFL to the point where it can fail spuriously.
I'm going to play with adding a message about the contents and some more robustness related to timing out.

**Comment by jonathanmetzman** [comment]
> > I'll try to reproduce this.
> > Does this target use threads by any chance?
> 
> Hi Jonathan,
> I'm having the same issue but my test cases are very simple.
> and the target is a multithreaded application.
> I used `-t +1000 `and I get
> 
> `[-] PROGRAM ABORT : All test cases time out, giving up!`
> 
> Is there any way to solve the problem?

@fatimahkj What project is this?

**Comment by asraa** [comment]
> . I bet because bad_build_check runs so many fuzzers in parallel it slows down AFL to the point where it can fail spuriously.

Ah! Okay good to know. There are definitely some fuzz tests in there at may run close to the 1000ms run, and I'll reduce those, and hope that that fixes the smaller broken fuzz tests.

**Comment by jonathanmetzman** [comment]
> > > I'll try to reproduce this.
> > > Does this target use threads by any chance?
> > 
> > 
> > Hi Jonathan,
> > I'm having the same issue but my test cases are very simple.
> > and the target is a multithreaded application.
> > I used `-t +1000 `and I get
> > `[-] PROGRAM ABORT : All test cases time out, giving up!`
> > Is there any way to solve the problem?
> 
> @fatimahkj What project is this?

If this isn't an issue you see in an OSS-Fuzz project (as https://github.com/google/AFL/issues/125 makes me think) then this isn't the correct place to discuss the problem.

**Comment by jonathanmetzman** [comment]
I think #4769 fixed this.
